### PR TITLE
GH #576: Make sure framebuffer attachments have  correct usage flag

### PR DESCRIPTION
**Pull request by mark-lunarg**

### CODE REVIEW

File: tests/layer_validation_tests.cpp [hunk 1]
```diff
@@ -2922,6 +2922,112 @@ TEST_F(VkLayerTest, BindMemoryToDestroyedObject) {
...
+    VkRenderPassCreateInfo rpci = {};
+    rpci.subpassCount = 1;
+    rpci.pSubpasses = &subpass;
+    rpci.attachmentCount = 1;
+    VkAttachmentDescription attach_desc = {};
+    attach_desc.format = VK_FORMAT_UNDEFINED;
+    rpci.pAttachments = &attach_desc;
+    rpci.sType = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO;
+    VkRenderPass rp;
+    vkCreateRenderPass(m_device->device(), &rpci, NULL, &rp);
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Error is generated here by your new check -- shouldn't be continuing beyond here.

File: tests/layer_validation_tests.cpp [hunk 2]
```diff
@@ -2922,6 +2922,112 @@ TEST_F(VkLayerTest, BindMemoryToDestroyedObject) {
...
+    fb_info.width = 100;
+    fb_info.height = 100;
+    fb_info.layers = 1;
+
+    VkFramebuffer fb;
+    VkResult result = vkCreateFramebuffer(device(), &fb_info, NULL, &fb);
+
+    m_errorMonitor->VerifyFound();
+
+    vkDestroyRenderPass(m_device->device(), rp, NULL);
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Tear these down in reverse order?

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -2922,6 +2922,112 @@ TEST_F(VkLayerTest, BindMemoryToDestroyedObject) {
...
+    VkRenderPassCreateInfo rpci = {};
+    rpci.subpassCount = 1;
+    rpci.pSubpasses = &subpass;
+    rpci.attachmentCount = 1;
+    VkAttachmentDescription attach_desc = {};
+    attach_desc.format = VK_FORMAT_UNDEFINED;
+    rpci.pAttachments = &attach_desc;
+    rpci.sType = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO;
+    VkRenderPass rp;
+    vkCreateRenderPass(m_device->device(), &rpci, NULL, &rp);
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Should probably assert that this succeeded. If it didn't, you'll get weird behavior later.

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -8668,42 +8724,50 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateRenderPass(VkDevice device, const VkRenderP
...
-        std::unordered_map<uint32_t, VkImageLayout> &attachment_first_layout =
-            render_pass->attachment_first_layout;
+        std::unordered_map<uint32_t, VkImageLayout> &attachment_first_layout = render_pass->attachment_first_layout;
         for (uint32_t i = 0; i < pCreateInfo->subpassCount; ++i) {
             const VkSubpassDescription &subpass = pCreateInfo->pSubpasses[i];
             if (subpass.pipelineBindPoint != VK_PIPELINE_BIND_POINT_GRAPHICS) {
-                skip_call |= log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, (VkDebugReportObjectTypeEXT)0, 0,
-                                     __LINE__, DRAWSTATE_INVALID_RENDERPASS, "DS",
-                                     "Pipeline bind point for subpass %d must be VK_PIPELINE_BIND_POINT_GRAPHICS.", i);
+                skip_call |=
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Not new, but this whole function is awful. Accumulating skip_call conditions after making the call!?

**Comment by mark-lunarg** [comment]
Thanks for the feedback @chrisforbes.  Changes incorporated and pushed to Master.  I'll create a new issue for the ugliness in CreateRenderPass.  

`commit 6600a7662f6dac6a8c8622a8305f02625439bf30
Author: Mark Lobodzinski <mark@lunarg.com>
Date:   Thu Jun 16 13:23:02 2016 -0600

    layers: GH576, Validate attachment image usage

    Framebuffer attachment images are created with IMAGE_USAGE flags. These
    flags are now validated against their attachment location in subpass
    description structures. Also added a check for a preserve attachment
    being set to UNUSED and fixed a few typos in the attachment index
    checking stuff.

    Change-Id: Id62d3539c490ce8f381d7ae606402980f11946fa`

**Pull request merged**